### ISSUE TITLE
Extract frames from animated HEIC/HEIF sequence images

### ISSUE THREAD

**Issue by lfcnassif** [sepinf-inc/IPED#854]
This is a spinoff of #763 and #852. Currently we support just GIFs and APNGs. Support could be extended for more formats, like HEIF and WEBP. And behavior for TIFs is different from GIFs/APNGs: TIFs have the first frame rendered in gallery and a specific viewer to navigate through the frames, while GIFs/APNGs have more than 1 frame displayed in a gallery cell and a frame mosaic is rendered in image viewer. I think displaying more than 1 frame in gallery is good, users would notice multiple frames images easily. Not sure what is the best approach in the viewer.

**Comment by tc-wleite** [comment]
@lfcnassif, I took a closer look about animated WEBP (mentioned in #852) and HEIC/HEIF sequences.
I would keep TIFF as it is today.
I think HEIC/HEIF sequences could already be handled by the current solution (MPlayer) and maybe WEBP decoding support in MPlayer will be available soon (https://trac.ffmpeg.org/ticket/4907).

Meanwhile, I would like to add HEIC/HEIF support, as sequences using these formats are somewhat common nowadays.
What do you think?

By the way, I noticed that HEIC/HEIF sequences are not receiving the "sequence" mime type.
I guess it should be "image/heic-sequence" and "image/heif-sequence" for those files.
The specification is under "HEIF Related Brands and MIME Types" in https://nokiatech.github.io/heif/technical.html.

I found this in **CustomSignatures.xml**:
```
	<mime-type type="image/heic">
    	<magic priority="60">
      		<match value="ftypheic" type="string" offset="4"/>
      		<match value="ftypheix" type="string" offset="4"/>
      		<match value="ftyphevc" type="string" offset="4"/>
      		<match value="ftyphevx" type="string" offset="4"/>
    	</magic>
    	<glob pattern="*.heic"/>
  	</mime-type>
  	
  	<mime-type type="image/heif">
    	<magic priority="60">
      		<match value="ftypmif1" type="string" offset="4"/>
      		<match value="ftypmsf1" type="string" offset="4"/>
    	</magic>
    	<glob pattern="*.heif"/>
  	</mime-type>
```

I noticed that even HEIF files (with "ftypmif1" and "ftypmsf1") are being detected as "image/heic" if their extension is ".heic".
How does "glob pattern" precedence work?
I thought the signature would always have a higher precedence, but it is not working as I expected, so I am probably misinterpreting this configuration file.

**Comment by lfcnassif** [comment]
> I would keep TIFF as it is today.

Actually I'm not sure what is the best approach for TIFs. Usually they are documents and displaying 1 page at a time in the viewer should be better for reading text, like is done for document types (PDFs, DOCs...). But maybe it's worth to display 2/3 pages for TIFs in the gallery like GIFs/APNGs, we can't read anything in the gallery today, although just first page of PDFs/DOCs is shown. Of course this could be targeted separatelly, when we have a final opinion.

> I think HEIC/HEIF sequences could already be handled by the current solution (MPlayer) and maybe WEBP decoding support in MPlayer will be available soon (https://trac.ffmpeg.org/ticket/4907).
> 
> Meanwhile, I would like to add HEIC/HEIF support, as sequences using these formats are somewhat common nowadays. What do you think?

Great, I think it's worthy!


> By the way, I noticed that HEIC/HEIF sequences are not receiving the "sequence" mime type.

This could be my fault. I'll read the specification later.

> I noticed that even HEIF files (with "ftypmif1" and "ftypmsf1") are being detected as "image/heic" if their extension is ".heic".
> How does "glob pattern" precedence work?

I think the problem is that heic is declared as sub-type of heif in Tika: https://github.com/apache/tika/blob/main/tika-core/src/main/resources/org/apache/tika/mime/tika-mimetypes.xml#L5681
If the signature matches, Tika detection rules use the extension to refine the mimetype. Maybe just overriding the `sub-class-of` in our custom definition file could fix this.

**Comment by tc-wleite** [comment]
> > Meanwhile, I would like to add HEIC/HEIF support, as sequences using these formats are somewhat common nowadays. What do you think?
> 
> Great, I think it's worthy!

Cool! I will try to do that.
 
> > By the way, I noticed that HEIC/HEIF sequences are not receiving the "sequence" mime type.
> 
> This could be my fault. I'll read the specification later.

Don't :-) 
I was just talking about this part:
```
Header  MIME Type
------  -------------------
mif1    image/heif
msf1    image/heif-sequence
heic    image/heic
heix    image/heic
hevc    image/heic-sequence
hevx    image/heic-sequence
```

> I think the problem is that heic is declared as sub-type of heif in Tika: https://github.com/apache/tika/blob/main/tika-core/src/main/resources/org/apache/tika/mime/tika-mimetypes.xml#L5681 If the signature matches, Tika detection rules use the extension to refine the mimetype. Maybe just overriding **edited**: `sub-class-of="application/octet-stream"` in our custom definition file could fix this.

Got it! Thanks!
I will take a look on this, as identifying which HEIC/HEIF are sequences is a prerequisite to handle them in a specific way.

**Comment by lfcnassif** [comment]
> This could be my fault. I'll read the specification later.

Actually, if I remember, I didn't care about the difference for *sequence mimes, as those would be handled equally until now, and put all signatures together.

**Comment by tc-wleite** [comment]
Some sample of HEIC/HEIF sequences:
[HEIC-HEIF-Sequences.zip](https://github.com/sepinf-inc/IPED/files/7634834/HEIC-HEIF-Sequences.zip)

**Comment by tc-wleite** [comment]
@lfcnassif, a couple of quick questions: 

1. Should I use this issue or create a new one (something like "Extract frames from HEIC/HEIF sequences").
2. What do you think of creating a subcategory ("Animated Images"), under "Images"?

**Comment by lfcnassif** [comment]
> 1. Should I use this issue or create a new one (something like "Extract frames from HEIC/HEIF sequences").

This issue could be used for HEIC/HEIF and I'll open another for WebP since it needs a future mplayer release or non trivial changes to ImageThumbTask.

> 2\. What do you think of creating a subcategory ("Animated Images"), under "Images"?

I think you already thought about that in #763, right? Since one item could belong to more than one category, I see no problem about overlapping with "Images in System Folders" or "Temporary Internet Images" categories. But non leaf categories item count would need to be changed, since it simply sum up the subcategories items count today...

**Comment by tc-wleite** [comment]
> > 1. Should I use this issue or create a new one (something like "Extract frames from HEIC/HEIF sequences").
> 
> This issue could be used for HEIC/HEIF and I'll open another for WebP since it needs a future mplayer release or non trivial changes to ImageThumbTask.

Great, thanks!

> > 2. What do you think of creating a subcategory ("Animated Images"), under "Images"?
> 
> I think you already thought about that in #763, right? Since one item could belong to more than one category, I see no problem about overlapping with "Images in System Folders" or "Temporary Internet Images" categories. But non leaf categories item count would need to be changed, since it simply sum up the subcategories items count today...

Correct! And I gave up because of this intersection...
Not sure if it is worth the effort.

**Comment by lfcnassif** [comment]
Closed by #858.
